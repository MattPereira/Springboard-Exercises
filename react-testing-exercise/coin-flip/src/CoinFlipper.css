.CoinFlipper {
    text-align: center;
}

.CoinFlipper-btn {
    background-color: dodgerblue;
    padding: 1.25rem;
    border-radius: 20px;
    color:white;
    font-size: 2rem;
    border: none;
    cursor: pointer;
    font-family: monospace;
    font-weight: 700;
    margin: 3rem;
}

button:hover{
    background-color:rgb(0, 128, 255);
}

.CoinFlipper-counter {
    color: white;
    font-family: monospace;
    font-size: 1.5rem;
}
