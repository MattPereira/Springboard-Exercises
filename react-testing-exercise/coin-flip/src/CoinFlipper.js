import React, { useState } from "react";
import tail from "./pennyTail.png";
import head from "./pennyHead.png";
import Coin from "./Coin";
import choice from "./helpers";
import "./CoinFlipper.css";

const CoinFlipper = (props) => {
  const [coinSide, setCoinSide] = useState(null);
  const [headsCount, setHeadsCount] = useState(0);
  const [tailsCount, setTailsCount] = useState(0);

  const handleClick = () => {
    const newSide = choice(props.sides);

    setCoinSide(newSide);
    if (newSide.name === "heads") {
      setHeadsCount(headsCount + 1);
    } else {
      setTailsCount(tailsCount + 1);
    }
  };

  const currSide = coinSide ? (
    <Coin imgSrc={coinSide.imgSrc} altText={coinSide.name} />
  ) : null;

  return (
    <div className="CoinFlipper">
      {currSide}
      <button className="CoinFlipper-btn" onClick={handleClick}>
        FLIP COIN
      </button>
      <h3 className="CoinFlipper-counter">
        Out of {headsCount + tailsCount} flips, there have been {headsCount}{" "}
        heads and {tailsCount} tails!
      </h3>
    </div>
  );
};

CoinFlipper.defaultProps = {
  sides: [
    {
      imgSrc: head,
      name: "heads",
    },
    {
      imgSrc: tail,
      name: "tails",
    },
  ],
};

export default CoinFlipper;
