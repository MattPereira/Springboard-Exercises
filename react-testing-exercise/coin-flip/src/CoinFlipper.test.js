import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import CoinFlipper from "./CoinFlipper";

//mock the return value of Math.random
beforeEach(function () {
  jest
    .spyOn(Math, "random")
    .mockReturnValueOnce(0.25)
    .mockReturnValueOnce(0.75);
});

/* SMOKE TEST*/
it("renders without crashing", function () {
  render(<CoinFlipper />);
});

/* SNAPSHOT TEST*/
it("matches snapshot", function () {
  const { asFragment } = render(<CoinFlipper />);
  expect(asFragment()).toMatchSnapshot();
});

/* SPECIALIZED TESTING*/
it("shows no coin image on page load", function () {
  render(<CoinFlipper />);
  expect(screen.queryByTestId("coin")).toBeNull();
});

it("shows coin when button clicked", function () {
  render(<CoinFlipper />);
  const btn = screen.queryByText("FLIP COIN");
  fireEvent.click(btn);

  expect(screen.getByTestId("coin")).not.toBeNull();
});

it("counts correctly when heads", function () {
  render(<CoinFlipper />);

  const btn = screen.queryByText("FLIP COIN");
  fireEvent.click(btn);

  expect(screen.getByAltText("heads")).toBeInTheDocument();
  expect(screen.queryByAltText("tails")).not.toBeInTheDocument();

  expect(
    screen.getByText("Out of 1 flips, there have been 1 heads and 0 tails!")
  ).toBeInTheDocument();
});

it("counts correctly when tails", function () {
  render(<CoinFlipper />);

  const btn = screen.queryByText("FLIP COIN");
  fireEvent.click(btn);
  fireEvent.click(btn);

  expect(screen.queryByAltText("heads")).not.toBeInTheDocument();
  expect(screen.getByAltText("tails")).toBeInTheDocument();

  expect(
    screen.getByText("Out of 2 flips, there have been 1 heads and 1 tails!")
  ).toBeInTheDocument();
});

//resets the mock return values of Math.random
afterEach(function () {
  Math.random.mockRestore();
});
