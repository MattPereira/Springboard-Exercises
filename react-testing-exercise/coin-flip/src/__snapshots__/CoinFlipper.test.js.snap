// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CoinFlipper"
  >
    <button
      class="CoinFlipper-btn"
    >
      FLIP COIN
    </button>
    <h3
      class="CoinFlipper-counter"
    >
      Out of 0 flips, there have been 0 heads and 0 tails!
    </h3>
  </div>
</DocumentFragment>
`;
